File: 958-check-completeness-of-a-binary-tree/958-check-completeness-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCompleteTree(TreeNode* root) {
        if ( root == nullptr)
            return true;
        
        std::queue<pair<TreeNode *, long long>>  q;
        
        q.push( pair<TreeNode *, long long>(root, 1));
        
        int expectedId = 1;
        
        while( !q.empty() ) {
            int lSize = q.size();
            
            for( int i = 0; i < lSize; i++) 
            {
                pair<TreeNode *, long long> p = q.front();
                q.pop();
                
                if( expectedId == p.second )
                    expectedId += 1;
                else
                    return false;
                
                if( p.first->left )
                    q.push( pair<TreeNode *, long long>(p.first->left, p.second * 2));
                
                if( p.first->right )
                    q.push( pair<TreeNode *, long long>(p.first->right, p.second * 2 + 1)); 
            }
        }
        
        return true;
    }
};

File: 958-check-completeness-of-a-binary-tree/958-check-completeness-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isCompleteTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None: 
            return True
        
        q = collections.deque([(root, 1)])
        expectedId = 1
        
        while len(q) != 0:
            numNodes = len(q)
            
            for _ in range(numNodes):
                (node, id) = q.popleft()
                if expectedId == id:
                    expectedId += 1
                else:
                    return False
                
                if node.left is not None:
                   q.append((node.left, id*2))
                if node.right is not None:
                    q.append((node.right, id*2 + 1))
        
        return True
                
                
            
        